Sources/CucumberSwift/Runner/Cucumber.swift

```diff
@@ -177,100 +177,106 @@ import CucumberSwift_ObjC
             .map { Feature(with: $0, uri: uri) })
     }
 
-    func attachClosureToSteps(keyword: Step.Keyword? = nil, regex: String, callback: @escaping (([String], Step) throws -> Void), line: Int, file: StaticString) {
-        features
+    func executeFirstStep(keyword: Step.Keyword? = nil, matching: String) {
+        let firstMatchingStep = features
             .flatMap { $0.scenarios.flatMap { $0.steps } }
-            .filter { step -> Bool in
+            .first {step -> Bool in
                 if  let k = keyword,
                     step.keyword.contains(k) {
-                    return !step.match.matches(for: regex).isEmpty
+                    return step.matchesExpression?(matching) == true
                 } else if keyword == nil {
-                    return !step.match.matches(for: regex).isEmpty
+                    return step.matchesExpression?(matching) == true
                 }
                 return false
             }
-            .forEach { step in
-                step.result = .undefined
-                step.execute = { try callback(step.match.matches(for: step.regex), step) }
-                step.regex = regex
-                step.sourceLine = line
-                step.sourceFile = file
-            }
+
+        if let firstMatchingStep = firstMatchingStep {
+            XCTAssertNoThrow(try firstMatchingStep.execute?(matching, firstMatchingStep))
```

Comment:
I'm gonna go ahead and merge. Might do a follow-up to add the line/file to this function and pass it to the XCTAssert and/or XCTFail just so failures are a little more traceable. 

I also wonder if we should check that `execute` isn't nil and throw an error if it is.